benchmark/grouping.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2022 The GPflow Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Code for grouping data for plotting.
+"""
+import heapq
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Collection, Iterator, List, Sequence, Tuple, Type
+
+import pandas as pd
+
+from benchmark.metadata import BenchmarkMetadata
+from benchmark.plotter_api import Plotter
+
+
+class GroupingKeySource(Enum):
+    """ Sources of data to group by. """
+
+    METADATA = "metadata"
+    METRICS_DF = "metrics_df"
+    PLOTTER = "plotter"
+
+
+class GroupingKey(Enum):
```

Comment:
Maybe include the member value type in the class definition? (mypy won't check it for you but it won't complain either)

```suggestion
class GroupingKey(Tuple[GroupingKeySource, type, bool, float], Enum):
```